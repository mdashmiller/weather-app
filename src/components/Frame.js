import styled from "styled-components"

const Frame = styled.div`
	width: 300px;
	margin: 40px auto 30px;
	text-align: center;
	font-family: 'Varela Round', Arial, sans-serif;
`

export default Frame
