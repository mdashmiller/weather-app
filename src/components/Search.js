import styled, { css } from "styled-components"

const Search = styled.input`
	background: #eee;
	border: 1px solid #c3c3c3;
	border-radius: 10px;
	margin: 200px auto 20px;
	width: 200px;
	height: 34px;
	padding-left: 8px;
	font-family: 'Font Awesome 5 Free', 'Varela Round', Arial, sans-serif;
	font-weight: 900;
	font-size: 18px;

	&::placeholder {
		color: #6a6a6a;
	}

	${props => props.short && css`
	    width: 114px;
	    padding-left: 5px;
	    margin: 0;
	  `}

	${props => props.landingPage && css`
		display: block;
	    width: 172px;
	   	margin: 26px auto;
		text-align: center;
		padding: 0;
		cursor: pointer;  
	  `}

	${props => props.userLocation && css`
		border: 3px solid #f46403;
			
			&:hover {
				background: #f46403;
				
				&::placeholder {
					color: #eee;
				}
			}	
	  `}

	${props => props.userZip && css`
		border: 3px solid #6a6a6a;

			&:hover {
				background: #6a6a6a;
				
				&::placeholder {
					color: #eee;
				}
			} 
	  `}
`

export default Search
