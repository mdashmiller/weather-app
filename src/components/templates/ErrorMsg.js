import React from 'react'

const ErrorMsg = () =>
	<div>
		<h2>It looks like openweathermap.org is having some issues.</h2>
		<h2>Please try again in a few minutes.</h2>
	</div>

export default ErrorMsg
