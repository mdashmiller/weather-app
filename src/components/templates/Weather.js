import React, { Component } from 'react'
import WeatherIcon from '../ui/WeatherIcon'
import ThermoIcon from '../ui/ThermoIcon'
import PropTypes from 'prop-types'

class Weather extends Component {

	state = {
		day: null,
		place: '',
		temp: '',
		description: '', 
		codition: null
	}

	// component methods

	setWeatherInfo = result => {
		// takes results from API call and
		// routes data to the proper channels
		this.setTemp(result.main.temp)
		this.setDescription(result.weather[0].description)
		this.setPlace(result.name)
		this.setCondition(result.weather[0].id)
		this.dayOrNight(result.dt, result.sys.sunrise, result.sys.sunset)
	}

	setTemp = tempK => {
		// sets this.state.temp to the result
		// obtained from Open Weather Map
		const tempF =  this.convertToFahrenheit(tempK)
		this.setState( {temp: tempF} )
	}

	convertToFahrenheit = k => {
		// takes a temperature in Kelvin
		// and returns it in Fahrenheit
		return (
			(1.8 * (k - 273)) + 32
		).toFixed(1)
	}

	setDescription = weather =>
		// sets this.state.description to the result
		// obtained from Open Weather Map
		this.setState( {description: weather })

	setPlace = place =>
		// sets this.state.name to the result
		// obtained from Open Weather Map
		this.setState( {place} )

	setCondition = id =>
		// sets this.state.condition to the result
		// obtained from Open Weather Map
		this.setState( {condition: id} )

	dayOrNight = (time, sunrise, sunset) => {
		// determines if it is day or night
		// and sets state accordingly
		if (time >= sunrise && time < sunset) {
			this.setState({ day: true })
		} else {
			this.setState({ day: false })
		}
	}

	// lifecycle hooks

	componentDidMount () {
		this.setWeatherInfo(this.props.result)
	}

	render () {
		const {
			day,
			place,
			temp,
			description,
			condition
		} = this.state
		return (
			<div>
				<h1 className="title-gold">
					{place}
				</h1>
				<h2>{temp} F</h2>
				<h3>{description}</h3>
				<WeatherIcon 
					code={condition} 
					day={day}
				/>
				<ThermoIcon temp={temp} />
			</div>
		)
	}
}

Weather.propTypes = {
	result: PropTypes.shape({
		name: PropTypes.string,
		main: PropTypes.shape({
			temp: PropTypes.number
		}),
		weather: PropTypes.arrayOf(
			PropTypes.shape({
				description: PropTypes.string,
				id: PropTypes.number
			})
		),
		dt: PropTypes.number,
		sys: PropTypes.shape({
			sunrise: PropTypes.number,
			sunset: PropTypes.number
		})
	})
}

export default Weather
