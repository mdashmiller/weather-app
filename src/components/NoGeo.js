import React from 'react'
import { Link } from 'react-router-dom'
import Frame from './Frame'
import Search from './Search'

const NoGeo = () =>
	<Frame>
		<Link to='/'>
			<h1 id="landing-title" className="title-gold">
				Weather
				<span className="title-grey">
					Now
				</span>
			</h1>
		</Link>
		<h2>It looks like geolocation is disabled.</h2>
		<h2>Would you like to search for weather by zipcode?</h2>
		<Link to='/lookup-by-zip'>
			<Search
				type="text"
				placeholder="&#xf002; Lookup By Zip"
				noGeo
				lookUp
			/>
		</Link>
	</Frame>

export default NoGeo
