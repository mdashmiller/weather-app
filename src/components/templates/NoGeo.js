import React from 'react'
import { Link } from 'react-router-dom'
import Frame from '../ui/Frame'
import Title from '../ui/Title'
import Search from '../ui/Search'

const NoGeo = () =>
	<Frame>
		<Title />
		<h2>It looks like geolocation is disabled.</h2>
		<h2>Would you like to search for weather by zipcode?</h2>
		<Link to='/lookup-by-zip'>
			<Search
				type="text"
				placeholder="&#xf002; Lookup By Zip"
				noGeo
				lookUp
			/>
		</Link>
	</Frame>

export default NoGeo
