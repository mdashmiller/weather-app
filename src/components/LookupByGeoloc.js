import React, { Component } from 'react'
import config from '../config'
import { Link } from 'react-router-dom'
import Canvas from './Canvas'
import Frame from './Frame'
import Weather from './Weather'
import Search from  './Search'

// Open Weather Map API url details
const PATH_BASE = 'http://api.openweathermap.org/data/2.5/weather?'
const KEY = config.key

class LookupByGeoloc extends Component {

	state = {
		lat: '',
		lon: '',
		result: null,
		error: null,
		day: null,
		noGeoLocation: false
	}

	// component methods

	getLocation = () => {
		// finds user's latitude and longitude
		if (navigator.geolocation) {
			navigator.geolocation.getCurrentPosition(position => {
				this.setCoords(position)
			})
		} else {
			this.setState({ noGeoLocation: true })
		}
	}

	setCoords = position => {
		// sets users lattitude and longitude  
		// in state and initiates API call
		this.setState({
			lat: position.coords.latitude,
			lon: position.coords.longitude
		})
		this.getWeather()
	}

	getWeather = () => {
		// call weather API with coordinates
		const COORDS = `lat=${this.state.lat}&lon=${this.state.lon}`
		const urlCoords = `${PATH_BASE}${COORDS}&APPID=${KEY}`
		fetch(urlCoords)
			.then(response => response.json())
			.then(result => this.setState({ result }))
			.catch(error => this.setState({ error }))
		this.state.result &&
			this.dayOrNight(
				this.state.result.dt,
				this.state.result.sys.sunrise,
				this.state.result.sys.sunset
			)
	}

	dayOrNight = (time, sunrise, sunset) => {
		// determines if it is day or night
		// and sets state accordingly
		if (time >= sunrise && time < sunset) {
			this.setState({ day: true })
		} else {
			this.setState({ day: false })
		}
	}

	// lifecycle hooks

	componentDidMount () {
		this.getLocation()
	}

	render () {
		const {
			result,
			day,
			error
		} = this.state
		if (result) {
			return (
				<Canvas
					className={ day ? 'day-bg' : 'night-bg' }
				>
					<Frame>
						<Weather
							result={result}
						/>
						<Link to='/lookup-by-zip'>
							<Search
								type="text"
								placeholder="&#xf002; Change Location"
								lookUp
							/>
						</Link>
					</Frame>
				</Canvas>
			)
		} else if (error) {
			return (
				<Canvas>
					<Frame>
						<p>{error.message}</p>
					</Frame>
				</Canvas>
			)
		} else {
			return (
				<Canvas>
					<Frame>
						<h1>Getting Weather...</h1>
					</Frame>
				</Canvas>
			)
		}
	}
}

export default LookupByGeoloc
