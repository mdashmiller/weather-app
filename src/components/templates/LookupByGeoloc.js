import React, { Component } from 'react'
import config from '../../config'
import { Link } from 'react-router-dom'
import Weather from './Weather'
import NoGeo from  './NoGeo'
import ErrorMsg from './ErrorMsg'
import Frame from '../ui/Frame'
import Search from  '../ui/Search'
import Title from '../ui/Title'

// Open Weather Map API url details
const PATH_BASE = 'http://api.openweathermap.org/data/2.5/weather?'
const KEY = config.key

class LookupByGeoloc extends Component {

	state = {
		lat: null,
		lon: null,
		result: {},
		error: null,
		noGeoLocation: false
	}

	// component methods

	getLocation = () => //{
		// finds user's latitude and longitude
		navigator.geolocation.getCurrentPosition(this.setCoords, this.geoError)

	setCoords = position => {
		// sets users lattitude and longitude  
		// in state and initiates API call
		this.setState({
			lat: position.coords.latitude,
			lon: position.coords.longitude
		})
		this.getWeather()
	}

	geoError = () =>
		// sets state to render a conditonal template
		// when users geolocation is disabled
		this.setState({ noGeoLocation: true })

	getWeather = () => {
		// call weather API with coordinates
		const COORDS = `lat=${this.state.lat}&lon=${this.state.lon}`
		const urlCoords = `${PATH_BASE}${COORDS}&APPID=${KEY}`
		fetch(urlCoords)
			.then(response => response.json())
			.then(result => this.setWeatherInfo(result))
			.catch(error => this.setState({ error }))
		
	}

	setWeatherInfo = result => 
		this.setState({ result })

	// lifecycle hooks

	componentDidMount () {
		this.getLocation()
	}

	render () {
		const {
			result,
			noGeoLocation,
			error
		} = this.state
		if (result.name) {
			return (
				<Frame>
					<Weather
						result={result}
					/>
					<Link to='/lookup-by-zip'>
						<Search
							type="text"
							placeholder="&#xf002; Change Location"
							lookUp
						/>
					</Link>
				</Frame>
			)
		} else if (noGeoLocation) {
			return (
				<NoGeo />
			)
		} else if (error) {
			return (		
				<Frame>
					<Title />
					<ErrorMsg />	    	
					<Search
						type="text"
						placeholder="&#xf3c5;  Try Again?"
						onClick={() => window.location.reload()}
						landingPage
						userLocation
						short
						geoError
					/>
				</Frame>				
			)		
		} else {
			return (			
				<Frame>
					<h2 className="geoloc-load-title">Getting Weather...</h2>
					<i className="fas fa-spinner fa-2x geoloc-spinner">
					</i>
				</Frame>			
			)
		}
	}
}

export default LookupByGeoloc
