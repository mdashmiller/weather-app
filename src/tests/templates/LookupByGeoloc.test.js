import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Enzyme, { shallow, render } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { BrowserRouter, Link } from 'react-router-dom'

import LookupByGeoloc from '../../components/templates/LookupByGeoloc'
import Frame from '../../components/ui/Frame'
import Title from '../../components/ui/Title'
import ErrorMsg from '../../components/templates/ErrorMsg'
import Search from '../../components/ui/Search'
import NoGeo from '../../components/templates/NoGeo'
import Weather from '../../components/templates/Weather'

Enzyme.configure({ adapter: new Adapter() })

describe('<LookupByGeoloc /> basic rendering', () => {

	it('renders without crashing', done => {
	  	const div = document.createElement('div')
	  	ReactDOM.render(
	  		<BrowserRouter>
	  			<LookupByGeoloc />
	  		</BrowserRouter>,
	  	div)

	  	setTimeout(() => {
	  		ReactDOM.unmountComponentAtNode(div)
	  		done()
	  	})
	  	
	})

	test('has a valid snapshot', () => {
		const component = renderer.create(
			<BrowserRouter>
				<LookupByGeoloc />
			</BrowserRouter>
		)
		let tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})

})

describe('<LookupByGeoloc /> rendering with no weather data', () => {

	let wrapper

	beforeAll(() => {
		wrapper = shallow(<LookupByGeoloc />)
	})

	beforeEach(() => {
		wrapper.setState({ result: { name: '' } })
	})

	it('renders 1 <Frame> component', () => {
		expect(wrapper.find(Frame).length).toBe(1)
	})

	it('renders 1 <h2> element', () => {
		expect(wrapper.find('h2').length).toBe(1)
	})

	it('renders 1 <i> element', () => {
		expect(wrapper.find('i').length).toBe(1)
	})

})

describe('<LookupByGeoloc /> rendering with an error', () => {

	let wrapper

	beforeAll(() => {
		wrapper = shallow(<LookupByGeoloc />)
	})

	beforeEach(() => {
		wrapper.setState({
			result: {
				name: ''
			},
			error: { 
				message: 'error' 
			}
		})
	})

	it('renders 1 <Frame> component', () => {
		expect(wrapper.find(Frame).length).toBe(1)
	})

	it('renders 1 <Title> component', () => {
		expect(wrapper.find(Title).length).toBe(1)
	})

	it('renders 1 <ErrorMsg> component', () => {
		expect(wrapper.find(ErrorMsg).length).toBe(1)
	})

	it('renders 1 <Search> component', () => {
		expect(wrapper.find(Search).length).toBe(1)
	})

})

describe('<LookupByGeoloc /> rendering with geolocation disabled', () => {

	it('renders 1 <NoGeo> component', () => {
		const wrapper = shallow(<LookupByGeoloc />)
		wrapper.setState({ noGeoLocation: true })

		expect(wrapper.find(NoGeo).length).toBe(1)
	})	

})

describe('<LookupByGeoloc /> rendering with weather data', () => {

	let wrapper

	beforeAll(() => {
		wrapper = shallow(<LookupByGeoloc />)
	})

	it('renders 1 <Frame> component', () => {
		expect(wrapper.find(Frame).length).toBe(1)
	})	

	it('renders 1 <Weather> component', () => {
		expect(wrapper.find(Weather).length).toBe(1)
	})

	it('renders 1 <Search> component', () => {
		expect(wrapper.find(Search).length).toBe(1)
	})	

	it('renders 1 <Link>', () => {
		expect(wrapper.find(Link).length).toBe(1)
	})

})

describe('<LookupByGeoloc> Link functionality when there is weather data', () => {

	it('contains a <Link> to get weather by zip', done => {
		const wrapper = shallow(<LookupByGeoloc />)

		setTimeout(() => {
			wrapper.update()

			expect(wrapper.find(Link).prop('to')).toBe('/lookup-by-zip')

			done()
		})
	})

})

describe('<LookupByGeoloc> Search functionality when there is an error', () => {

	it('reloads the page when clicked', () => {
		const wrapper = shallow(<LookupByGeoloc />)
		wrapper.setState({
			error: {
				message: 'error'
			}
		})
		window.location.reload = jest.fn()

		wrapper.find(Search).simulate('click')
		
		expect(window.location.reload).toHaveBeenCalled()
	})

})

describe('weather data fetched on mount with geolocation enabled', () => {

	it('sets the API result in state', done => {
		const wrapper = shallow(<LookupByGeoloc />)

		setTimeout(() => {
			wrapper.update()
			const state = wrapper.instance().state

			expect(state.result.name).toBe('Mockville')
			expect(state.result.main.temp).toBe(297)
			expect(state.result.weather[0].description).toBe('all good!')
			expect(state.result.weather[0].id).toBe(100)
			expect(state.result.dt).toBe(2)
			expect(state.result.sys.sunrise).toBe(1)
			expect(state.result.sys.sunset).toBe(3)

			done()
		})

	})

})

// describe('<NoGeo> displayed on mount with geolocation disabled', () => {

// 	jest.mock('geolocationForceError')

// 	it('sets noGeoLocation in state and renders <NoGeo> template', done => {
// 		const wrapper = shallow(<LookupByGeoloc />)

// 		setTimeout(() => {
// 			wrapper.update()
// 			const state = wrapper.instance().state

// 			expect(state.noGeoLocation).toBe(true)
// 			expect(wrapper.find(NoGeo).length).toBe(1)

// 			done()
// 		})
// 	})

// })
