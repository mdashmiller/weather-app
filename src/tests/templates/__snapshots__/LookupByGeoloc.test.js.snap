// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LookupByGeoloc /> basic rendering has a valid snapshot 1`] = `
<div
  className="sc-bwzfXH kUBsGr"
>
  <h2
    className="geoloc-load-title"
  >
    Getting Weather...
  </h2>
  <i
    className="fas fa-spinner fa-2x geoloc-spinner"
  />
</div>
`;
